### PR TITLE
`mps`: Alternative implementation for `repeat_interleave`

### LINKED ISSUE #760: StableDiffusionPipeline producing unexpected output with MPS device using diffusers==0.4.0
### Describe the bug

I tried testing the potential speed updates of diffusers 0.4.0 on my M1 mac using an existing StableDiffusionPipeline-based script, and I found that a large image that would take ~3 min to generate in diffusers 0.3.0 was estimated to take more than 10x as long.

Since my existing script had a lot going on (e.g. large resolutions, attention slicing), I tried to diagnose the problem with a minimal script (see below), running in two identical environments, with the only difference being the diffusers version. 

In diffusers 0.3.0, it takes ~35 seconds to generate a reasonable result like this:
test_0 3 0

In diffusers 0.4.0, it takes ~50 seconds (which is slower than 0.3.0, but better that the 10x performance hit I was getting before), but each attempt (with varying seeds) triggered the NSFW filter. Disabling the filter, the results appear to be just noise:
test_0 4 0

I'm not sure if the initial 10x performance hit I initially observed in my original script would be fixed by fixing this bug, but this certainly seems to be at least part of it. 

### Reproduction

import torch
from diffusers import StableDiffusionPipeline

pipe = StableDiffusionPipeline.from_pretrained("CompVis/stable-diffusion-v1-4")
pipe.to("mps")

result = pipe("dogs playing poker", generator=torch.manual_seed(1))

result.images[0].save("test.png")

### Logs

```shell
Under 0.4.0 there's also this warning:

/opt/homebrew/Caskroom/miniforge/base/envs/sd/lib/python3.10/site-packages/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:222: UserWarning: The operator 'aten::repeat_interleave.self_int' is not currently supported on the MPS backend and will fall back to run on the CPU. This may have performance implications. (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/mps/MPSFallback.mm:11.)
  text_embeddings = text_embeddings.repeat_interleave(num_images_per_prompt, dim=0)
```

### System Info

- `diffusers` version: 0.4.0
- Platform: macOS-12.6-arm64-arm-64bit
- Python version: 3.10.6
- PyTorch version (GPU?): 1.13.0.dev20220911 (False)
- Huggingface_hub version: 0.10.0
- Transformers version: 4.21.3
- Using GPU in script?: MPS
- Using distributed or parallel set-up in script?: no

### PR DESCRIPTION
**Pull request by pcuenca**
Fixes #760.

`repeat_interleave` was slowing things down, but not causing wrong outputs. Something between `v0.3.0` and `v0.4.0` (possibly some of the optimization changes) broke generation in `mps`. Fortunately, the latest `mps` version of PyTorch nightly (released today) works fine 😅, so I'm updating the documentation to recommend it.

### CODE REVIEW

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 1]
```diff
@@ -257,7 +264,14 @@ def __call__(
             uncond_embeddings = self.text_encoder(uncond_input.input_ids.to(self.device))[0]
 
             # duplicate unconditional embeddings for each generation per prompt
-            uncond_embeddings = uncond_embeddings.repeat_interleave(batch_size * num_images_per_prompt, dim=0)
+            if self.device.type == "mps":
+                # Workaround for `repeat_interleave`. Assumes 3 dims.
+                d0, d1, _ = uncond_embeddings.shape
+                uncond_embeddings = uncond_embeddings.repeat(1, num_images_per_prompt, 1).view(
+                    d0 * num_images_per_prompt, d1, -1
+                )
```
Comment (patil-suraj) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:272]:
```suggestion
            if self.device.type == "mps":
                # Workaround for `repeat_interleave`. Assumes 3 dims.
                 batch_size, seq_len, _ = uncond_embeddings.shape
                uncond_embeddings = uncond_embeddings.repeat(1, num_images_per_prompt, 1).view(
                    batch_size * num_images_per_prompt, seq_len, -1
                )
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 2]
```diff
@@ -257,7 +264,14 @@ def __call__(
             uncond_embeddings = self.text_encoder(uncond_input.input_ids.to(self.device))[0]
 
             # duplicate unconditional embeddings for each generation per prompt
-            uncond_embeddings = uncond_embeddings.repeat_interleave(batch_size * num_images_per_prompt, dim=0)
+            if self.device.type == "mps":
```
Comment (patil-suraj) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:267]: can we always use this, instead of checking for device here ?

**Review by patil-suraj — COMMENTED** [review]
Looks good, just left couple of comments.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:267]: +1 - I think this makes more sense

**Comment by pcuenca** [comment]
Changes applied, I think this is ready to go @patil-suraj @patrickvonplaten

**Comment by pcuenca** [comment]
Hmm, the test if failing, I'll verify it locally.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 3]
```diff
@@ -218,8 +218,11 @@ def __call__(
             text_input_ids = text_input_ids[:, : self.tokenizer.model_max_length]
         text_embeddings = self.text_encoder(text_input_ids.to(self.device))[0]
 
-        # duplicate text embeddings for each generation per prompt
-        text_embeddings = text_embeddings.repeat_interleave(num_images_per_prompt, dim=0)
+        # duplicate text embeddings for each generation per prompt, using mps friendly method
+        batch_size, seq_len, _ = text_embeddings.shape
+        text_embeddings = text_embeddings.repeat(1, num_images_per_prompt, 1).view(
+            batch_size * num_images_per_prompt, seq_len, -1
+        )
```
Comment (patil-suraj) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:225]: (nit) maybe split `repeat` and `view` in multiple lines for better readability.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 4]
```diff
@@ -256,8 +259,11 @@ def __call__(
             )
             uncond_embeddings = self.text_encoder(uncond_input.input_ids.to(self.device))[0]
 
-            # duplicate unconditional embeddings for each generation per prompt
-            uncond_embeddings = uncond_embeddings.repeat_interleave(batch_size * num_images_per_prompt, dim=0)
+            # duplicate unconditional embeddings for each generation per prompt, using mps friendly method
+            batch_size, seq_len, _ = uncond_embeddings.shape
+            uncond_embeddings = uncond_embeddings.repeat(1, num_images_per_prompt, 1).view(
+                batch_size * num_images_per_prompt, seq_len, -1
+            )
```
Comment (patil-suraj) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:266]: same comment as above

**Review by patil-suraj — APPROVED** [review]
Good to merge for me once the tests pass.

**Review by patrickvonplaten — APPROVED** [review]

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py (hunk 3 above)
Comment (pcuenca) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:225]: Good idea. I just applied what `make style` thought was ok.

**Pull request merged**